internal/jobworker/grpc/server.go

```diff
@@ -0,0 +1,230 @@
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// logger is an object for logging package events to stdout.
+var logger = log.New(os.Stdout, "grpc")
+
+// NewJobWorker creates a JobWorker instance.
+func NewJobWorker(jobSvc *job.Service, userSvc IUserService) *JobWorker {
+	return &JobWorker{jobSvc: jobSvc, userSvc: userSvc}
+}
+
+var _ pb.JobWorkerServiceServer = (*JobWorker)(nil)
+
+// IUserService provides an API for interacting with jobworker users.
+type IUserService interface {
+	// User retrieves the user associated with the ctx. The ok return value
+	// should indicate if the user could be retrieved. The user return value
+	// should be the user's unique identifer.
+	User(ctx context.Context) (string, bool)
+}
+
+// Jobworker provides mechanisms for starting, stopping, fetching status, and
+// output streaming jobs.
+// Jobworker implements pb.JobWorkerServiceServer.
+type JobWorker struct {
+	jobSvc  *job.Service
+	userSvc IUserService
+}
+
+func (jw JobWorker) Start(ctx context.Context, req *pb.StartRequest) (*pb.StartResponse, error) {
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	valid := validator.New()
+	valid.AssertFunc(func() bool { return req.Command != nil }, "command empty")
+	valid.AssertFunc(func() bool { return req.Command.Name != "" }, "command name empty")
+	valid.AssertFunc(func() bool { return req.Limits != nil }, "limits empty")
+	if err := valid.Err(); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	logger.Infof("processing StartRequest; Command: %v", req.Command)
+
+	j, err := job.New(
+		user,
+		reexec.Command{
+			Name: req.Command.Name,
+			Args: req.Command.Args,
+		},
+	)
+	if err != nil {
+		logger.Errorf("building Job; error: %v", err)
+		return nil, status.Error(codes.Internal, "error building job")
+	}
+
+	if err := jw.jobSvc.StartJob(
+		ctx,
+		*j,
+		cgroupOptions(req.Limits)...,
+	); err != nil {
+		logger.Errorf("starting Job; error: %v", err)
+		return nil, status.Error(codes.Internal, "error starting job")
+	}
+
+	logger.Infof("Job started; ID: %v", j.ID)
+	return &pb.StartResponse{
+		JobId:   j.ID.String(),
+		Command: req.Command,
+		Status: &pb.StatusDetail{
+			Status:   toStatus(j.Status()),
+			ExitCode: int32(j.ExitCode()),
+		},
+		Limits: req.Limits,
+	}, nil
+}
+
+func (jw JobWorker) Stop(ctx context.Context, req *pb.StopRequest) (*pb.StopResponse, error) {
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	if req.JobId == "" {
+		return nil, status.Error(codes.InvalidArgument, validator.Format("empty job ID"))
+	}
+
+	j, err := jw.fetchJob(ctx, user, req.JobId)
+	if err != nil {
+		return nil, err
+	}
+
+	if j.Status() != job.Running {
+		return nil, status.Error(codes.FailedPrecondition, "job is not running")
```

Comment:
I think we should roll this into the library, seems like it's a responsibility it should have.